core/finality-grandpa/src/import.rs

```diff
@@ -174,75 +168,69 @@ impl<'a, Block: 'a + BlockT> Drop for PendingSetChanges<'a, Block> {
 	}
 }
 
-impl<B, E, Block: BlockT<Hash=H256>, RA, PRA, SC>
-	GrandpaBlockImport<B, E, Block, RA, PRA, SC>
+fn find_scheduled_change<B: BlockT>(header: &B::Header)
+	-> Option<ScheduledChange<NumberFor<B>>>
+{
+	let id = OpaqueDigestItemId::Consensus(&GRANDPA_ENGINE_ID);
+
+	let filter_log = |log: ConsensusLog<NumberFor<B>>| match log {
+		ConsensusLog::ScheduledChange(change) => Some(change),
+		_ => None,
+	};
+
+	// find the first consensus digest with the right ID which converts to
+	// the right kind of consensus log.
+	header.digest().convert_first(|l| l.try_to(id).and_then(filter_log))
+}
+
+fn find_forced_change<B: BlockT>(header: &B::Header)
```

Comment:
We can obtain a marginal speedup by fusing this function with `find_scheduled_change`, but I doubt it is worthwhile.  We do, however, need to check that there is at most one of each kind of consensus digest, as I did in BABE.  Alternatively, if multiple such digests are allowed, they need to be handled appropriately.

Reply:
We give preference to the first digest; subsequent ones are allowed but ignored.